bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+			return err
+		}
+	}
+
+	// Create the table within the dataset.
+	tableSchema := getMetricsTableSchema()
+	fmt.Fprintln(w, "Metrics Table Schema:")
+	for _, field := range tableSchema {
+		fmt.Fprintf(w, "  %s (%s)\n", field.Name, field.Type)
+	}
+	fmt.Fprintln(w)
+
+	// NOTE: Not setting an expiration for BigQuery tables created by the endpoint.
+	// If an expiration was used and someone forgets to extend the deadline,
+	// the table and all of its data are deleted by BigQuery without warning.
+	metaData := &bq.TableMetadata{
+		Schema: tableSchema,
+		//ExpirationTime: time.Now().AddDate(1, 0, 0), // Table will be automatically deleted in 1 year
+	}
+	tableRef := session.client.Dataset(t.dataset).Table(t.tablename)
+	if err := tableRef.Create(session.ctx, metaData); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "already exist") {
+			return fmt.Errorf("Error creating table '%s': %s\n", t.formatTableId(), err)
+		}
+	} else {
+		fmt.Fprintf(w, "Created table %s\n", t.formatTableId())
+	}
+	return nil
+}
+
+// Translate protobuf metric to BigQuery metric.
+func translateMetric(stream bazelStream, m testMetric) (bigQueryMetric, error) {
+	// Process the slice of key/value pairs representing individual metric tags.
+	dat := make(map[string]string)
+	for _, kv := range m.tags {
+		dat[kv.key] = kv.value
+	}
+	// Insert additional tags using information that is only available
+	// to the BES endpoint through the Bazel event message itself.
+	//
+	// Storing both the invocationId and the invocationSha in the database,
+	// the former being useful to match up with build log references, etc.
+	dat["invocation_id"] = stream.invocationId
+	dat["invocation_sha"] = stream.invocationSha
+	dat["run"] = stream.run
+	dat["test_name"] = stream.testName
+	tags, err := json.Marshal(dat)
+	if err != nil {
+		return bigQueryMetric{}, fmt.Errorf("Error converting JSON to string: %s", err)
+	}
+
+	// Change the metric timestamp from integer to formatted string,
+	// since that is what BigQuery expects for a TIMESTAMP column.
+	//
+	// The incoming timestamp is UTC nanoseconds since epoch, but
+	// storing with microsecond granularity in BigQuery.
+	ts := m.timestamp / 1000
+	tsf := time.Unix(ts/1000000, ts%1000000).Format("2006-01-02 15:04:05.000000")
+
+	return bigQueryMetric{
+		metricname: m.metricname,
+		tags:       string(tags),
+		value:      m.value,
+		timestamp:  tsf,
+	}, nil
+}
+
+// Upload this set of metrics to the specified BigQuery table.
+func uploadTestMetrics(w io.Writer, stream bazelStream, r *metricTestResult) error {
+	// Normalize the BigQuery table identifier based on whether one was specified
+	// in the protobuf message.
+	normalizeTableRef(&r.table)
+	fmt.Fprintf(w, "Normalized table ref: %s\n", r.table.formatTableId())
+
+	// Get client context for this BigQuery operation.
+	ctx := context.Background()
+	client, err := bq.NewClient(ctx, r.table.project)
```

Comment:
This is currently failing in the staging cluster due to lack of auth credentials for accessing BigQuery. Adam is looking into a fix; will revisit this call once the change is determined.

Reply:
+1

Reply:
This has since been fixed by Adam. Confirmed this latest BES Endpoint code is able to push metrics into BigQuery from the staging cluster.